### PR TITLE
MeshPhysicalMaterial: Support iridescence / thin-film materials

### PR DESCRIPTION
**Pull request by PascalSchoen**
**Description**

Implementation of the KHR_materials_iridescence glTF extension.

### CODE REVIEW

**Comment by donmccurdy** [comment]
Preview Link:

- [Editor](http://raw.githack.com/PascalSchoen/three.js/feature/KHR_materials_iridescence/editor/index.html)

Screen Shot 2022-04-08 at 12 06 13 PM

**Comment by mrdoob** [comment]
Ops... the builds are already conflicting. Could you remove them from the PR?

**Comment by PascalSchoen** [comment]
> Ops... the builds are already conflicting. Could you remove them from the PR?

@mrdoob Yeah, didn't think of this conflict. Removed the build files from the PR. ;)

File: src/renderers/webgl/WebGLPrograms.js [hunk 1]
```diff
@@ -184,6 +185,10 @@ function WebGLPrograms( renderer, cubemaps, cubeuvmaps, extensions, capabilities
 			clearcoatRoughnessMap: useClearcoat && !! material.clearcoatRoughnessMap,
 			clearcoatNormalMap: useClearcoat && !! material.clearcoatNormalMap,
 
+			iridescence: useIridescence,
+			iridescenceMap: useIridescence && !! material.iridescenceMap,
+			iridescenceThicknessMap: useIridescence && !! material.iridescenceThicknessMap,
```
Comment (Mugen87) [inline src/renderers/webgl/WebGLPrograms.js:190]: You have to honor these new program parameters in `getProgramCacheKeyBooleans()`.

Comment (PascalSchoen) [reply src/renderers/webgl/WebGLPrograms.js:190]: @Mugen87 done ✅

**Review by Mugen87 — APPROVED** [review]

**Comment by elalish** [comment]
Thanks @PascalSchoen! Looks good to me, though I'd like to know what @WestLangley thinks. I'd still like to refactor multiscattering, since I don't think it's right that we make up the energy difference with the diffuse lobe. Would be nice to simplify the code there too, but that can happen separately from this PR.

File: src/materials/MeshPhysicalMaterial.js [hunk 2]
```diff
@@ -39,6 +39,12 @@ class MeshPhysicalMaterial extends MeshStandardMaterial {
 			}
 		} );
 
+		this.iridescenceMap = null;
+		this.iridescenceIOR = 1.3;
+		this.iridescenceThicknessMinimum = 100;
+		this.iridescenceThicknessMaximum = 400;
```
Comment (mrdoob) [inline src/materials/MeshPhysicalMaterial.js:45]: I think this is the first time we have these type of properties in the material.

Would it make sense to unify them into `iridescenceThicknessRange`? ie..

```js
this.iridescenceThicknessRange = [ 100, 400 ];
```

Comment (PascalSchoen) [reply src/materials/MeshPhysicalMaterial.js]: @mrdoob I changed the PR to use a range instead of separated values. Also updated the editor.

**Comment by mrdoob** [comment]
@PascalSchoen Lookin good! Is there any sample model we can use in the examples that uses this?

**Comment by emackey** [comment]
Just a heads-up, Khronos PBR TSG may need to change `"iridescenceIOR"` to `"iridescenceIor"` to fix the camelCase.

In the meantime, here are some existing samples using the old casing:

- [IridescenceDielectricSpheres](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/IridescenceDielectricSpheres)
- [IridescenceMetallicSpheres](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/IridescenceMetallicSpheres)  <- If you're tired of Suzanne, this one is the coolest.
- [IridescenceSuzanne](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/IridescenceSuzanne)

**Comment by takahirox** [comment]
[IridescenceMetallicSpheres](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/IridescenceMetallicSpheres) looks cool but looks like testing purpose too much? Personally I'm ok with it but if we don't really prefer testing purpose models, probably Suzanne may be the only option in them.

File: src/materials/MeshPhysicalMaterial.js [hunk 3]
```diff
@@ -40,7 +40,7 @@ class MeshPhysicalMaterial extends MeshStandardMaterial {
 		} );
 
 		this.iridescenceMap = null;
-		this.iridescenceIOR = 1.3;
```
Comment (donmccurdy) [inline src/materials/MeshPhysicalMaterial.js]: @PascalSchoen sorry for the churn here, but glTF's camelCase convention for acronyms like XYZ (`fooBarXyz`) is different from what we follow in three.js (`fooBarXYZ`). I think for the public API of MeshPhysicalMaterial, this should remain `iridescenceIOR`. Feel free to use whichever you prefer for internal code in the rest of the PR. 🙏

Comment (donmccurdy) [reply src/materials/MeshPhysicalMaterial.js]: I think so, yes — for example we use:

- attribute.getXYZ not .getXyz
- attribute.x not attribute.X
- GLTFLoader not GltfLoader

I'm comfortable with either convention but keeping acronyms uppercase seems to be the norm in three.js today, and (I think?) more common in JavaScript generally.

Comment (donmccurdy) [reply src/materials/MeshPhysicalMaterial.js]: Slight preference for applyIORToRoughness, yes - but not as worried about internal methods.

Somehow `.ior` does not bother me, feels similar to `vec.xyz` in GLSL. 🤷‍♂️

**Review by donmccurdy — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
Depending on the model the shader could crash because of `dotNV` redefinition.  https://github.com/mrdoob/three.js/commit/e60a8027463e04b0a6b060c3537798e599086c58

**Comment by emackey** [comment]
> IridescenceMetallicSpheres looks cool but looks like testing purpose too much?

@takahirox Khronos has a new photo-realistic model now from Wayfair:  [IridescenceLamp](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/IridescenceLamp).

**Comment by takahirox** [comment]
> @takahirox Khronos has a new photo-realistic model now from Wayfair: [IridescenceLamp](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/IridescenceLamp).

Thanks. That looks much nicer if we add an example for iridescence / thin-film materials.

**Comment by mrdoob** [comment]
Yes, I noticed. I was going to do an example today 😊

Screen Shot 2022-05-26 at 11 15 34 AM

**Pull request merged**